src/test/java/com/google/cloud/tools/managedcloudsdk/internal/download/StandardDownloaderTest.java

```diff
@@ -0,0 +1,174 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk.internal.download;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.URLStreamHandler;
+import java.nio.charset.Charset;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.NotDirectoryException;
+import java.nio.file.Path;
+import org.apache.commons.io.FileUtils;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+public class StandardDownloaderTest {
+
+  @Rule public TemporaryFolder tmp = new TemporaryFolder();
+
+  @Test
+  public void testValidateAndCreateFiles_targetDirectoryIsFile()
+      throws IOException, InterruptedException {
+    Path fileWithBadParent = tmp.newFile().toPath().resolve("xyz");
+
+    StandardDownloader downloader = new StandardDownloader(null, fileWithBadParent, null, null);
+    try {
+      downloader.call();
+      Assert.fail("NotDirectoryException expected but not thrown");
+    } catch (NotDirectoryException nde) {
+      // pass
+    }
+  }
+
+  @Test
+  public void testValidateAndCreateFiles_fileAlreadyExists()
+      throws IOException, InterruptedException {
+    Path fileThatExists = tmp.newFile().toPath();
+
+    StandardDownloader downloader = new StandardDownloader(null, fileThatExists, null, null);
+    try {
+      downloader.call();
+      Assert.fail("FileAlreadyExists expected but not thrown");
+    } catch (FileAlreadyExistsException faee) {
+      // pass
+    }
+  }
+
+  @Test
+  public void testDownloadURL_createsNewDirectory() throws IOException, InterruptedException {
+    Path destination = tmp.getRoot().toPath().resolve("dir-to-create").resolve("destination-file");
+    Path testSourceFile = createTestRemoteResource(1);
+    URL fakeRemoteResource = testSourceFile.toUri().toURL();
+
+    StandardDownloader downloader =
+        new StandardDownloader(fakeRemoteResource, destination, null, null);
+
+    downloader.call();
+  }
+
+  @Test
+  public void testDownloadURL_worksWithNullProgressListener()
+      throws IOException, InterruptedException {
+
+    Path destination = tmp.getRoot().toPath().resolve("destination-file");
+    Path testSourceFile = createTestRemoteResource(1);
+    URL fakeRemoteResource = testSourceFile.toUri().toURL();
+
+    StandardDownloader downloader =
+        new StandardDownloader(fakeRemoteResource, destination, null, null);
+
+    downloader.call();
+  }
+
+  @Test
+  public void testDownloadURL() throws IOException, InterruptedException {
+    Path destination = tmp.getRoot().toPath().resolve("destination-file");
+    Path testSourceFile = createTestRemoteResource(StandardDownloader.BUFFER_SIZE * 10);
```

Comment:
I think my original comment got outdated:

As mentioned in the StandardDownloader file, changing this to BUFFER_SIZE * 161 for example, breaks this test.